### PR TITLE
defining test matrix in tox and switchint to tox-travis

### PR DESCRIPTION
**Pull request by miki725**
this allows to nicely run complete matrix locally

```
  py27-dj11-sqla11: commands succeeded
  py27-dj11-sqla12: commands succeeded
  py27-dj11-sqla13: commands succeeded
  py36-dj11-sqla11: commands succeeded
  py36-dj11-sqla12: commands succeeded
  py36-dj11-sqla13: commands succeeded
  py36-dj20-sqla11: commands succeeded
  py36-dj20-sqla12: commands succeeded
  py36-dj20-sqla13: commands succeeded
  py36-dj21-sqla11: commands succeeded
  py36-dj21-sqla12: commands succeeded
  py36-dj21-sqla13: commands succeeded
  py36-dj22-sqla11: commands succeeded
  py36-dj22-sqla12: commands succeeded
  py36-dj22-sqla13: commands succeeded
  py37-dj20-sqla11: commands succeeded
  py37-dj20-sqla12: commands succeeded
  py37-dj20-sqla13: commands succeeded
  py37-dj21-sqla11: commands succeeded
  py37-dj21-sqla12: commands succeeded
  py37-dj21-sqla13: commands succeeded
  py37-dj22-sqla11: commands succeeded
  py37-dj22-sqla12: commands succeeded
  py37-dj22-sqla13: commands succeeded
  congratulations :)
```

### CODE REVIEW

**Comment by miki725** [comment]
pypy fails for me for now with psycopg stuff. might be that im linking to c dep not compatible with pypy like openssl1.1 or something system level:

```
ERROR: Command "/Users/miki725/Code/miki725/django-sorcery/.tox/pypy-dj11-sqla11/bin/pypy -u -c 'import setuptools, tokenize;__file__='"'"'/private/var/
folders/4z/bxmdspnn5qs9x0bbz8w7cbrw0000gn/T/pip-install-2aStt0/psycopg2/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().
replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/4z/bxmdspnn5qs9x0bb
z8w7cbrw0000gn/T/pip-record-E95CVK/install-record.txt --single-version-externally-managed --compile --install-headers /Users/miki725/Code/miki725/django
-sorcery/.tox/pypy-dj11-sqla11/include/site/python2.7/psycopg2" failed with error code 1 in /private/var/folders/4z/bxmdspnn5qs9x0bbz8w7cbrw0000gn/T/pip
-install-2aStt0/psycopg2/
```

but eventually more python interpreters can be added fairly easily to tox. cool thing that even though travis now uses tox but it only runs interpreters specified in travis config file which means tox could define larger matrix and travis wont execute pypy for example since historically travis is a bit tricky with pypy especially with pypy3

**Comment by miki725** [comment]
let me know if you like this. I can send something similar to witchcraft

File: tox.ini [hunk 1]
```diff
@@ -1,15 +1,39 @@
...
+    pypy3: pypy3
+deps =
+    dj11: django<2
+    dj20: django<2.1
+    dj21: django<2.2
+    dj22: django<2.3
+    sqla11: sqlalchemy<1.2
+    sqla12: sqlalchemy<1.3
+    sqla13: sqlalchemy<1.4
+    pipenv
```
Comment (shosca) [inline tox.ini]: drop

File: .travis.yml [hunk 2]
```diff
@@ -46,12 +18,10 @@ before_script:
   - make -j resetdb
```
Comment (shosca) [inline .travis.yml]: probably don't need this anymore

Comment (miki725) [reply .travis.yml]: good catch

File: tox.ini [hunk 3]
```diff
@@ -1,15 +1,44 @@
...
+whitelist_externals =
+    make
+commands =
+    pip freeze
+    make -j resetdb
     make coverage
+
+[testenv:py37-dj22-sqla13]
+setenv =
+    COVERAGE_FLAGS = --cov-fail-under=100
```
Comment (miki725) [inline tox.ini]: pypy did not play nice with coverage stuff. according to local coverage report it was blank lines sometimes in docstring. could be related to docstring tests so skipping coverage minimum by default except a single py3 build where I also run lint stuff

**Comment by miki725** [comment]
with pypy and pypy3 added:

```
  py27-dj11-sqla11: commands succeeded
  py27-dj11-sqla12: commands succeeded
  py27-dj11-sqla13: commands succeeded
  py36-dj11-sqla11: commands succeeded
  py36-dj11-sqla12: commands succeeded
  py36-dj11-sqla13: commands succeeded
  pypy-dj11-sqla11: commands succeeded
  pypy-dj11-sqla12: commands succeeded
  pypy-dj11-sqla13: commands succeeded
  py36-dj20-sqla11: commands succeeded
  py36-dj20-sqla12: commands succeeded
  py36-dj20-sqla13: commands succeeded
  py36-dj21-sqla11: commands succeeded
  py36-dj21-sqla12: commands succeeded
  py36-dj21-sqla13: commands succeeded
  py36-dj22-sqla11: commands succeeded
  py36-dj22-sqla12: commands succeeded
  py36-dj22-sqla13: commands succeeded
  py37-dj20-sqla11: commands succeeded
  py37-dj20-sqla12: commands succeeded
  py37-dj20-sqla13: commands succeeded
  py37-dj21-sqla11: commands succeeded
  py37-dj21-sqla12: commands succeeded
  py37-dj21-sqla13: commands succeeded
  py37-dj22-sqla11: commands succeeded
  py37-dj22-sqla12: commands succeeded
  py37-dj22-sqla13: commands succeeded
  pypy3-dj20-sqla11: commands succeeded
  pypy3-dj20-sqla12: commands succeeded
  pypy3-dj20-sqla13: commands succeeded
  pypy3-dj21-sqla11: commands succeeded
  pypy3-dj21-sqla12: commands succeeded
  pypy3-dj21-sqla13: commands succeeded
  pypy3-dj22-sqla11: commands succeeded
  pypy3-dj22-sqla12: commands succeeded
  pypy3-dj22-sqla13: commands succeeded
  congratulations :)
```

**Comment by miki725** [comment]
@shosca another 👀 ?

File: requirements.txt [hunk 4]
```diff
@@ -9,12 +9,13 @@ gitchangelog
 mock; python_version <= '2.7'
 pdbpp
 pre_commit
-psycopg2
```
Comment (miki725) [inline requirements.txt]: not installable on pypy but cffi variant is api compatible and seems to work everywhere. should be good enough for testing

**Pull request merged**